[nnc] Support thread level parallelism in fused kernels

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63545
* #63514
* __->__ #63386

Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

bertmaher committed:
[nnc] Support thread level parallelism in fused kernels

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30360382).

bertmaher committed:
Update on "[nnc] Support thread level parallelism in fused kernels"


Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Support thread level parallelism in fused kernels"


Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Support thread level parallelism in fused kernels"


Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30360382).

bertmaher committed:
Update on "[nnc] Support thread level parallelism in fused kernels"


Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30360382).

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2487,6 +2488,78 @@ void fuseAllLoops(StmtPtr st) {
+  for (int64_t i = loops.size(); i > 0; i--) {
+    auto const& loop = loops[i - 1];
+    if (auto stop = to<IntImm>(loop->stop())) {
+      grainSize *= stop->value();
```
This is assuming the loops are normalized at this point. While that might be true mostly, I'm not sure that is guaranteed.

bertmaher replied:
I don't think anything's likely to un-normalize the loops before this point, but maybe I should simplify (and do stop-start) just to be safe.  I guess the worst that happens is we miss a parallelization opportunity.

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2487,6 +2488,78 @@ void fuseAllLoops(StmtPtr st) {
+template <typename Bufs>
+static void parallelizeOuterLoops(LoopNest& l, Bufs&& bufs) {
+  for (auto const& buf : bufs) {
+    auto loops = l.getLoopStmtsFor(buf);
```
Since this function is called after `fuseAllLoops` it is possible that multiple buffers belong to the same loopnest. So, we could be repeating this loop multiple times for the same loopnest. I understand that that may not be incorrect at this point, but it could lead to bugs in future.

IMO, we shouldn't be looking at output buffers and their loopnests. Instead, we should just take `root_stmt` in the given `LoopNest` and apply parallelization for all loopnests in that stmt. Wdyt?

bertmaher replied:
That seems OK to me, sure.

bertmaher replied:
I guess where things could get a little weird is if multiple buffers are updated at different levels of the loopnest, e.g.:
```
for i:
  y1[] = ...
  for j:
    y2[] = ...
```

The current approach sort of gives each buffer an "independent" chance to affect the loop parallelization.  That doesn't seem terrible, tbh.

bertmaher replied:
Actually the more I think about it, is there really any advantage to starting with the root stmt and working down?  From my POV it just makes the code a lot more complicated; the way things happen now I just get a nice vector of loops leading to a buffer and try to flatten them.  If it's not flattenable, it simply fails and I give up.

Although, maybe it's not too much work to build up my own vector starting from the root.  Idk.

navahgar replied:
If you are okay with having the same set of loops being handled here for different bufs, then I have no objections to it. 

Personally, I felt starting from the root_stmt might be better. We might need another API to extract all loops in the root_stmt. So, may be we can do this in future.

huiguoo replied:
If I understand correctly,  going through all buffers would not miss parallelism opportunities like the following
```
for i 
  for j1
      y1 = ... [data dependence exists between iterations]
  for j2
     y2 = ... [no data dependence between iterations]
``` 
`i+j1` cannot be parallelized because there's data dependence between iterations for `y1`; but `i+j2` can be parallelized and we should not miss it. If this is the thing we try to do here, I guess we need a `distribute` transformation before `flatten`. `flatten` currently only handles perfectly nested loops.

bertmaher replied:
Yeah the approach here will definitely miss opportunities where some nested loops are parallelizable.  It's really kind of a best-effort thing to get simple elementwise fusions right, not a general solution to parallelism.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2487,6 +2488,78 @@ void fuseAllLoops(StmtPtr st) {
+      continue;
+    }
+    // Try to flatten the outer loops and parallelize them if successful.
+    For* flattened = nullptr;
```
Nit: ForPtr please ;)

navahgar reviewed (approved):
LGTM

huiguoo commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -274,15 +274,24 @@ class LLVMCodeGenImpl : public IRVisitor {
+        callee(index, packed_data);
+      }
+    });
+  } catch (...) {
```
kinda curious about this place: why not terminate if there's an exception? I guess the execution of the left stmts would ultimately lead to wrong results？

bertmaher replied:
Interesting point... if an exception happens here things are really screwed up, because we don't know how to unwind past llvm-generated frames.  But no exceptions should be possible here, since we're just parallel-dispatching to our own kernel, which doesn't throw exceptions.  So I was mainly putting the try-catch here to ensure that the compiler knew that it wouldn't need to unwind this frame.

bertmaher committed:
Update on "[nnc] Support thread level parallelism in fused kernels"


Differential Revision: [D30360382](https://our.internmc.facebook.com/intern/diff/D30360382)

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30360382).
